### PR TITLE
[dbnode] Support cold index writes

### PR DESCRIPTION
**Pull request by notbdu**
**What this PR does / why we need it**:
This change adds the ability to write out indexed series written arbitrarily out of order in the past (cold writes). Based off of work done in this PR: https://github.com/m3db/m3/pull/1825. 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/storage/index.go [hunk 1]
```diff
@@ -740,11 +746,26 @@ func (i *nsIndex) Tick(c context.Cancellable, startTime time.Time) (namespaceInd
 	return result, multiErr.FinalError()
 }
 
-func (i *nsIndex) Flush(
+func (i *nsIndex) WarmFlush(
 	flush persist.IndexFlush,
 	shards []databaseShard,
 ) error {
-	flushable, err := i.flushableBlocks(shards)
+	return i.flush(flush, shards, series.WarmWrite)
```
Comment (richardartoul) [inline src/dbnode/storage/index.go]: hmm I don't think this is what that `series.WarmWrite` constant is supposed to be used for. I'd prefer you just passed a bool or made your own constant than overload this one which is used on the write path and has a very specific meaning (unless we've already established this pattern somewhere which is possible)

**Comment by anandsinghkunwar** [comment]
What is the status of this PR?

**Comment by lz006** [comment]
what do you guys need to fulfill this "ColdWrite" feature? need a helping hand?

**Comment by notbdu** [comment]
We're in process of testing/reviewing.

File: src/dbnode/storage/types.go [hunk 2]
```diff
@@ -354,6 +354,11 @@ type databaseNamespace interface {
 		flush persist.FlushPreparer,
```
Comment (anandsinghkunwar) [inline src/dbnode/storage/types.go]: Shouldn't `FlushIndex` be renamed as well?

**Comment by asafm** [comment]
@notbdu I saw you closed this PR and curios where is the work continue? Is it in the original PR? (#1825 which is still open?)

**Pull request closed without merging**